File: api/categories.php
<?php
/**
 * inphub: expense categories CRUD.
 *
 *   GET  ?action=list
 *   POST ?action=create { name, color?, icon?, monthly_budget? }
 *   POST ?action=update { id, name?, color?, icon?, monthly_budget? }
 *   POST ?action=delete { id }   (expenses keep, category_id set NULL)
 */

require_once __DIR__ . '/_bootstrap.php';

api_handle(function (): void {
    $uid   = current_user_id();
    $input = request_input();

    switch (action($input)) {
        case 'list': {
            $stmt = db()->prepare('SELECT * FROM expense_categories WHERE user_id = ? ORDER BY name ASC');
            $stmt->execute([$uid]);
            ok($stmt->fetchAll());
            break;
        }

        case 'create': {
            $name = str_or_null(input_get($input, 'name'));
            if ($name === null) {
                fail('Category name is required.', 422);
            }
            try {
                $stmt = db()->prepare(
                    'INSERT INTO expense_categories (user_id, name, color, icon, monthly_budget)
                     VALUES (?, ?, ?, ?, ?)'
                );
                $stmt->execute([
                    $uid,
                    $name,
                    str_or_null(input_get($input, 'color')) ?? '#6b7280',
                    str_or_null(input_get($input, 'icon')),
                    num_or_null(input_get($input, 'monthly_budget')),
                ]);
            } catch (PDOException $e) {
                if ($e->getCode() === '23000') {
                    fail('A category with that name already exists.', 409);
                }
                throw $e;
            }
            ok(['id' => (int) db()->lastInsertId()]);
            break;
        }

        case 'update': {
            $cat = fetch_owned('expense_categories', (int) input_get($input, 'id'), $uid);
            $stmt = db()->prepare(
                'UPDATE expense_categories SET name=?, color=?, icon=?, monthly_budget=? WHERE id=? AND user_id=?'
            );
            $stmt->execute([
                str_or_null(input_get($input, 'name', $cat['name'])) ?? $cat['name'],
                str_or_null(input_get($input, 'color', $cat['color'])) ?? $cat['color'],
                array_key_exists('icon', $input) ? str_or_null($input['icon']) : $cat['icon'],
                array_key_exists('monthly_budget', $input) ? num_or_null($input['monthly_budget']) : $cat['monthly_budget'],
                (int) $cat['id'],
                $uid,
            ]);
            ok(['id' => (int) $cat['id']]);
            break;
        }

        case 'delete': {
            $cat = fetch_owned('expense_categories', (int) input_get($input, 'id'), $uid);
            $del = db()->prepare('DELETE FROM expense_categories WHERE id=? AND user_id=?');
            $del->execute([(int) $cat['id'], $uid]);
            ok(['deleted' => (int) $cat['id']]);
            break;
        }

        default:
            fail('Unknown action.', 404);
    }
});

File: server/Endpoints/Categories.cs
using Inphub.Core;
using MySqlConnector;

namespace Inphub.Endpoints;

public static class Categories
{
    // GET/POST /api/categories: list, create, update, delete (expenses keep, category set to NULL).
    public static void Map(WebApplication app) =>
        app.MapMethods("/api/categories", ["GET", "POST"], Api.Handle(Run));

    // Dispatches on ?action=.
    static object? Run(Req req)
    {
        var uid = req.Uid;
        var input = req.Input;

        switch (req.Action)
        {
            case "list":
                return Db.Rows("SELECT * FROM expense_categories WHERE user_id = @uid ORDER BY name ASC", new { uid });
            case "create":
            {
                var name = input.Str("name") ?? throw Api.Fail("Category name is required.", 422);
                try
                {
                    var id = Db.Insert(
                        @"INSERT INTO expense_categories (user_id, name, color, icon, monthly_budget)
                          VALUES (@uid, @name, @color, @icon, @budget)",
                        new
                        {
                            uid, name,
                            color = input.Str("color") ?? "#6b7280",
                            icon = input.Str("icon"),
                            budget = input.NumOrNull("monthly_budget"),
                        });
                    return new { id };
                }
                catch (MySqlException e) when (e.ErrorCode == MySqlErrorCode.DuplicateKeyEntry)
                {
                    throw Api.Fail("A category with that name already exists.", 409);
                }
            }
            case "update":
            {
                var cat = Api.FetchOwned("expense_categories", input.Int("id"), uid);
                var id = Convert.ToInt32(cat["id"]);
                Db.Execute(
                    "UPDATE expense_categories SET name = @name, color = @color, icon = @icon, monthly_budget = @budget WHERE id = @id AND user_id = @uid",
                    new
                    {
                        id, uid,
                        name = input.Str("name") ?? (string)cat["name"]!,
                        color = input.Str("color") ?? (string)cat["color"]!,
                        icon = input.StrIfSent("icon", cat["icon"]),
                        budget = input.NumIfSent("monthly_budget", cat["monthly_budget"]),
                    });
                return new { id };
            }
            case "delete":
            {
                var cat = Api.FetchOwned("expense_categories", input.Int("id"), uid);
                var id = Convert.ToInt32(cat["id"]);
                Db.Execute("DELETE FROM expense_categories WHERE id = @id AND user_id = @uid", new { id, uid });
                return new { deleted = id };
            }
            default:
                throw Api.Fail("Unknown action.", 404);
        }
    }
}
